specific/cmo_sale_discount_total_enhance/models/sale.py

```diff
@@ -27,32 +31,42 @@ class sale_order(models.Model):
         track_visibility='always',
     )
 
-    @api.model
+    @api.multi
     def calculate_discount(self, vals):
+        self.ensure_one()
         discount = 0
+        last_line = False
         if ('discount_rate' in vals) and ('order_line' in vals):
             res = vals['sale_order_record']
+            cur = res.pricelist_id.currency_id
             if vals['discount_type'] == 'percent':
                 discount = vals['discount_rate']
+                order_lines = res.order_line
             else:
                 amount = sum(res.order_line.mapped(
                     lambda r: r.product_uom_qty * r.price_unit)
                 )
                 total = amount if amount != 0 else 1 # prevent devison by zero
                 discount_rate = vals['discount_rate']
-                if discount_rate != 0:
-                    discount = (discount_rate / total) * 100.0
-                else:
-                    discount = discount_rate
-            for line in res.order_line:
+                discount = round((discount_rate / total) * 100.0, 16)
```

Comment:
Using float_round() is better than round()
Note: For some magic number, standard round() of python is not correct according to accounting rule.